File: packages/lang-python/templates/graph.py
"""The campus as the player walks it.

Two jobs: give the player something ergonomic to explore, and count. Every call to
``neighbors`` is one node expanded, and that is the metric the game shows instead of time
(ADR 0011).
"""

from dataclasses import dataclass

_explored = 0


def reset_ops() -> None:
    global _explored
    _explored = 0


def ops_count() -> int:
    return _explored


@dataclass(frozen=True)
class Edge:
    """One path leaving a building, with the distance to walk it."""

    to: str
    weight: float


class Graph:
    def __init__(self, raw: dict) -> None:
        self._order: list[str] = []
        self._labels: dict[str, str] = {}
        self._adjacency: dict[str, list[Edge]] = {}

        for node in raw.get("nodes", []):
            self._order.append(node["id"])
            self._labels[node["id"]] = node["label"]
            self._adjacency[node["id"]] = []

        # Edges are undirected and declared once, so both directions are filled in here.
        for edge in raw.get("edges", []):
            self._adjacency.setdefault(edge["from"], []).append(Edge(edge["to"], edge["weight"]))
            self._adjacency.setdefault(edge["to"], []).append(Edge(edge["from"], edge["weight"]))

    def nodes(self) -> list[str]:
        """Every node id in the graph."""
        return self._order

    def neighbors(self, id: str) -> list[Edge]:
        """The edges leaving a node. Each call counts as one node explored."""
        global _explored
        _explored += 1
        return self._adjacency.get(id, [])

    def label(self, id: str) -> str:
        """The name a person reads, as opposed to the id an algorithm uses."""
        return self._labels.get(id, id)

File: packages/lang-go/templates/graph.go
package main

import "encoding/json"

// Edge is one path leaving a building, with the distance to walk it.
type Edge struct {
	To     string  `json:"to"`
	Weight float64 `json:"weight"`
}

// Graph is the campus as the player walks it.
//
// It has two jobs: give the player something ergonomic to explore, and count. Every call to
// Neighbors is one node expanded, and that is the metric the game shows instead of time
// (ADR 0011).
type Graph struct {
	order     []string
	labels    map[string]string
	adjacency map[string][]Edge
}

var explored int

func resetOps() {
	explored = 0
}

func opsCount() int {
	return explored
}

// Nodes lists every node id in the graph.
func (g Graph) Nodes() []string {
	return g.order
}

// Neighbors gives the edges leaving a node. Each call counts as one node explored.
func (g Graph) Neighbors(id string) []Edge {
	explored++
	return g.adjacency[id]
}

// Label is the name a person reads, as opposed to the id an algorithm uses.
func (g Graph) Label(id string) string {
	if name, found := g.labels[id]; found {
		return name
	}
	return id
}

type rawGraph struct {
	Nodes []struct {
		ID    string `json:"id"`
		Label string `json:"label"`
	} `json:"nodes"`
	Edges []struct {
		From   string  `json:"from"`
		To     string  `json:"to"`
		Weight float64 `json:"weight"`
	} `json:"edges"`
}

func buildGraph(raw json.RawMessage) (Graph, error) {
	var source rawGraph
	if err := json.Unmarshal(raw, &source); err != nil {
		return Graph{}, err
	}

	graph := Graph{
		order:     make([]string, 0, len(source.Nodes)),
		labels:    make(map[string]string, len(source.Nodes)),
		adjacency: make(map[string][]Edge, len(source.Nodes)),
	}

	for _, node := range source.Nodes {
		graph.order = append(graph.order, node.ID)
		graph.labels[node.ID] = node.Label
		graph.adjacency[node.ID] = nil
	}

	// Edges are undirected and declared once, so both directions are filled in here.
	for _, edge := range source.Edges {
		graph.adjacency[edge.From] = append(graph.adjacency[edge.From], Edge{To: edge.To, Weight: edge.Weight})
		graph.adjacency[edge.To] = append(graph.adjacency[edge.To], Edge{To: edge.From, Weight: edge.Weight})
	}

	return graph, nil
}
